### ISSUE TITLE
Training UBM issue

### ISSUE THREAD

**Issue by code-player** [ppwwyyxx/speaker-recognition#38]
What does function 'get_all_data_fpaths()' in file speaker-recognition/tree/master/src/testbench/train-ubm.py is expected to read from 'test-data/mfcc-lpc-data/' (even directory doesn't exist)?
Is it some '.mfcc-lpc' file? If so, what does it contains?
Can't train UBM on my machine due to this issue.

**Comment by ppwwyyxx** [comment]
UBM code are not expected to work and is not used in either command line or GUI.
#32

**Comment by code-player** [comment]
I need to train UBM model for the new speakers, so executing that script file directly through command line. I know it's never used anywhere in your project. But I need to predict users along with the rejection too. So, please guide me for the same.

**Comment by ppwwyyxx** [comment]
UBM is just a giant GMM. You can just train it with `sklearn.mixture.GaussianMixture`.

**Comment by code-player** [comment]
Thats same as simple GMM Model. Then how is it used to predict the acceptance or rejection of results by GMM Model in case of unenrolled user.?

**Comment by ppwwyyxx** [comment]
A voice get rejected if its score given by the best person-specific model is not substantially higher than the score given by ubm.

**Comment by code-player** [comment]
I get that but how do I get those results. Could u just explain me that person specific model?

**Comment by code-player** [comment]
How can I generate those models?

**Comment by ppwwyyxx** [comment]
To do speaker recognition you'll train a GMM for each speaker.
UBM is a GMM trained on voices of different speakers.

**Comment by code-player** [comment]
If UBM is GMM trained on voices of different speakers , how is it different from the one generated during enrollment process in the project? I thought that was generated during enrollment.

**Comment by ppwwyyxx** [comment]
In enrollment you'll train a GMM for each speaker.

**Comment by code-player** [comment]
I think I'm not getting it clearly. How those 2 are different?

**Comment by ppwwyyxx** [comment]
In enrollment you'll train N GMMs for N speakers.

**Comment by code-player** [comment]
And in other case..?

**Comment by ppwwyyxx** [comment]
In other case you train 1 GMM for N speakers

**Comment by code-player** [comment]
And also there are files like 'ubm.mixture-32.utt-300.model' in src/gui/model directory which contains some numeric array data which is used for user prediction in the project itself. What does that data signify?

**Comment by ppwwyyxx** [comment]
This is used to initialize the training of per-person GMM.
It is generated by training a GMM on all the data we have.

**Comment by code-player** [comment]
So, are those the MFCC and LPC extracted features of all the available users or what ? Because in case it's a GMM object it can't be transformed to an array type, I think.

**Comment by ppwwyyxx** [comment]
yes

**Comment by chrisspen** [comment]
If a UBM is just a GMM trained on all the audio samples, should we be able to train our own UBM by essentially just enrolling each speech sample under a fake "ubm" label, and then pickling that GMM instance as our "ubm.model"?

**Comment by ppwwyyxx** [comment]
yes

**Comment by chrisspen** [comment]
If that's the case, why is there a separate UBM class in UBM.py? Does the class GMMSet expect an instance of the UBM class passed in the `ubm` parameter, or does it expect an instance of the GMM class?

**Comment by ppwwyyxx** [comment]
It expects an instance of the GMM class.

**Comment by chrisspen** [comment]
Thanks. Inside the `task_enroll` function, I added another call to `enroll` like:

    m.enroll(UBM_LABEL, fs, signal)

and then I modified `task_predict` to show nothing if it returns UBM_LABEL, and that seems to detect "unknown" samples fairly well.

**Comment by chrisspen** [comment]
Also, thanks for publishing this project. Of all the speaker recognition systems I've tried, this seems to be one of the faster and more accurate ones. Great job.
